!BA Edits required to have working Main branch. Uses sample FPS project as simple base.

Personwithhat opened the pull request:
The goal of this PR is to be able to launch a small sample project using main branch code.  
No new features (aside from sample project), all bug fixes.  
As few edits as possible, very clearly split commits, yada yada.  
Much cleaner than my upcoming Time edits, mostly cuz this is easier to track.  
	
Some things wont build. SDL_MIXER, prefab plugin, etc. so I disabled them.  
Aside from possibly sub-par 'fixes', at least the bugs and causes are isolated.

Two bugs/issues that I've yet to solve/work-around:  
1) The current 5.5 template level assets are incompatible with the main branch due to this [commit](https://github.com/CRYTEK/CRYENGINE/commit/6c921f1de5b68b355f7dcf7b4710e412c2af8e40).
		
	  The series of changes based around converting the chunk'd flag from 32 to 64 bit.  ["uint32 m_dwRndFlags;" to "uint64 m_dwRndFlags;"](https://github.com/CRYTEK/CRYENGINE/blob/6c921f1de5b68b355f7dcf7b4710e412c2af8e40/Code/CryEngine/Cry3DEngine/ObjectsTree_Serialize.cpp#L29)
		   
	Simply changing objectstree_serialize.cpp to handle 32bit again (while breaking other samples no doubt) lets you boot up the `example` map of the standard FPS template project.   
I don't know how to convert older asset files to newer format, and I don't have any 'new' working samples I can use as a test project.
		   
2) [_Exception thrown at 0x00007FFCCB2CA388 in GameLauncher.exe: Microsoft C++ exception: \_com_error at memory location 0x000000854339D210._](https://github.com/CRYTEK/CRYENGINE/blob/main/Code/CryEngine/RenderDll/XRenderD3D9/D3DSystem.cpp#L2158		)  
com_error is normal. D3D things. Not in a location that looks like it can/will do anything lol.


I'm available on slack for Q's/discussion, typically german hours since I do night-shifts :P 
Been talking with Thomas Wollenzin. I appreciate his help and responsiveness!

P.S.>> Sorry for any formatting issues and the title edits. Wish I could hide them.

SDKs can be downloaded [here](https://github.com/Personwithhat/CRYENGINE/releases/tag/5.5.0_PLUS)

Personwithhat commented:
Out of interest, does anyone in CE even look at PR's?

Does anyone outside of community managers that say "Devs spend their own time on stuff sorry" go on forums/Slack?

I can't reach anyone from CE in any channel that I try and it's going on 2 months. Quite frankly this is absurd. Not even the bare minimum of a conversation piece/starter.........

Cry-Flare commented:
Developers have, do and will check PRs. Time constraints among other factors can delay the review of these Pull Requests but it is something we are actively working on.

We appreciate you taking the time to submit this request but we must decline this PR in its current state. Please also be aware that main is a volatile branch and sometimes will have some issues.

https://github.com/CRYTEK/CRYENGINE/pull/510/commits/0835c21a22116595219c8cbb6e6dcb150c5462f9 and https://github.com/CRYTEK/CRYENGINE/pull/510/commits/31999c2e2662a03d61d15dc2f43c4ac207969995
We cannot accept “hacks” and fixes that address the symptoms and not the cause. If the reason for the issue is unknown then more investigation is required and should be made the resolve the issue at the source.

https://github.com/CRYTEK/CRYENGINE/pull/510/commits/00358d542af65679a8a928b4fc14596e8b2d668d
No crash was observed from our side. While it is true that erase() invalidates iterators, and a commit to address just that can be accepted, in reality when erasing the last element the iterator to the last element will point to end(). Doing a std::iter_swap doesn’t have any benefit (it is a vector, specifically a vector of pointers and copying a POD is identical to moving it, if anything std::iter_swap is slower) over just erase()-ing the last element. The loop invariant ensures that ‘b’ is valid and it will terminate, correctly, when we erase the penultimate iterator (i.e. the last element).

The suggestion is create small PRs that focus on a single feature, or single bug fix. The less separate commits and files touched, the quicker and faster the review and integration if possible.

Personwithhat commented:
#### Regarding 00358d5

I've already replied to the commit owner's comment via email, he'd pinged me and mentioned similar reasons, (but didn't get back to me). Overall I don't put fixes where they shouldn't be. My 'Main' PR is as minimal as possible, if I did it then it's because the engine didn't build or crashed on run-time.

I've tested this on two computers, clean install and what not. Ofc, I run my engine/sandbox/etc. all in Debug-build (not profile or release) so that might be the reason this isn't being caught. I've had issues in debug-only builds before with CE. 

Trust me when I say I know how to maintain a project, at least enough to let it build out of the box + have clear records of what packages/etc. need to be maintained and so on. Clean install of a computer -> final product steps are always something I worry about. CE could use a little of this sooner rather than later with its SDK's/general compilation setup. (I improved on the current system a bit in my new-standard branch). Regardless, each of my bugs is not on a whim and tested as much as I could with my limited Game-dev knowledge. I mean, it's why I caught (and solved!) so many to begin with!


#### PR's
This PR, as with all my PR's, are focused primarily on bringing issues up to CryTek's attention along with possible solutions. Typically a commit is a complete solution, with a few rare exceptions. They may be sub-optimal, but they fulfill their purpose well enough. It's better than simply adding an 'issue' with minimal detail. Since I don't know multi-threading/shader-compilation some 'full' solutions are too far out of my experience.

I don't expect any PR's to pass instantly or even in the state that they were pushed......that's rarely the case in any company. After all, even my 'complete' solutions might be judged sub-par or not conforming to some standard, hence requiring rewrites or optimizations. Happens all the time. However, these fixes/bug-traces save hours of your valuable time, so why not spare a few minutes on review? Quite possibly you **do**, but it's not visible to me or to many old-timer's in the CE community (AFAIK).

I can, of course, submit 23 PR requests. That sounds hilarious/counter-intuitive, and if one PR takes 1 month to reject based on very 'generic' reasons/stuff-seen-on-first-glance, then I doubt 23 PR's will get anywhere. I mean, the root reason is "Main branch wont even build, much less run", so I decided to group it all in a cohesive segment. Commit-specific comments can be done anyways for discussion purposes and all the edits as a whole can be tested easily.

If you do want to just test one portion of this, why not simply cherry-pick? It's harder to make sure 20-30 edits work together than each one individually (although ofc I made sure of both here).

Lemme know if you really want me to split each commit here into a separate PR..........I can do that ofc, and cross-post a 'main branch issues' post on your forums or something. Up to you.


#### But......
Unfortunately, this is (yet again) a community manager saying that "Devs check PRs". If they check PR's then it'd be swell if they could say why they can't add it and what not (e.g. we don't like hacky fixes even if they are to just show an issue).

And as much as I appreciate the comment, it would have been nice to get it a **month** ago........

With all the hours of work I'm putting in, it just feels to me that as a company CE does not return a corresponding olive-branch. Any messages to me are drastically delayed, unless I specifically mention CE's behavior (which can be viewed as aggressive/slander, and I intend neither). I'm stressing this too much, but it took a month to get a reply that could've been sent within 5 minutes of **any** dev's time.

I've cleaned up my commits, isolated core CE issues, rewritten and rewritten plenty of my edits to save as much of your time as possible in solving **your** bugs.........and I've yet to have a normal (or even nominal) discussion with anyone in charge of PR QA/development. This isn't even talking about my MASSIVE time-rewrite, which is an undertaking that may take 1-2 months of testing and correspondence!!

Please, in order to develop a proper working relationship between CE and the community as a whole, it would be in your (and my!) best interests to foster a stable communication channel. Please be responsive! I ofc understand scheduling issues, crunch-time pushes, and so on, but there definitely people responsible for this already. Taking a few minutes out of their day to post (at least) a quick "Checking this out" or "We don't like hacky fixes" would go far in improving visibility!

As it stands, you have next to no PR's and scant few issues (they've mostly accumulated over several months....). Perhaps when the flow gets too large your current stance would make more sense, but as a 3'd party I do not see the point! You may be having a catastrophic code-crunch meltdown but all **I** see is generic turn-arounds and poor response times. I can read every forum message over the past month in 2-3 hours tops. Every PR for the past half a year in an hour or two. 

This isn't even considering improving documentation (especially networking), actively answering questions (although there are few on CE forums) regarding specific code, and so on........

If this wall of text is anything to go by, then it shows that I truly care about working with CE and improving the engine. I'm not here to argue, or point fingers, I just want to take what CE has and improve on it. Let bygones be bygones, but without )at bare minimum) improved communications I do not see a future for CE and the community.

Cry-Flare commented:
The reasons given came from developers that looked into this pull request. If you find specific issues and are unsure of the correct way to fix it the proper method is to submit an issue.

The pull request was denied as a whole due to parts of the PR not being compatible with our standards and quality which was explained in the previous comment.

Please keep on topic to the PR in question to help keep the pull requests as easy as possible to process. If you have feedback regarding processes you are welcome to contact us via email at contact [at] cryengine [dot] com.
